### ISSUE TITLE
[tt-train] Parameter frozen with set_requires_grad(false) keeps being updated by every optimizer

### ISSUE THREAD

**Issue by Patriot117** [tenstorrent/tt-metal#57751]
### Component / Area

tt-train (optimizers)

### Issue Type (optional)

Bad Outputs

### Observed

If a parameter is already registered with a tt-train optimizer and then gets frozen with `set_requires_grad(false)`, every later `optimizer.step()` keeps updating it with the last gradient it received before the freeze. With SGD momentum, the momentum buffer keeps absorbing that same gradient as well. torch.optim leaves a frozen parameter unchanged.

This affects every C++ optimizer I checked: AdamW, AdamWFullPrecision, MorehAdamW, AdamWComposite, SGD, SGDComposite and MuonComposite. There's no error or warning.

You hit it with the common "freeze some layers partway through training" pattern when the optimizer was built over all of the model's parameters, for example:

```python
optimizer = ttml.optimizers.AdamW(model.parameters(), cfg)
# ... train for a while ...
for name, p in model.parameters().items():
    if name.startswith("embedding"):
        p.set_requires_grad(False)
# ... keep training: the embedding still moves every step
```

### Cause (main at c87cf141a5)

Three things combine:

1. `Tensor::set_requires_grad` only flips the flag and keeps `m_grad` (`tt-train/sources/ttml/autograd/tensor.cpp:135-137`).
2. `Tensor::add_grad` returns early when `m_requires_grad` is false (`tensor.cpp:35-39`), so backward never overwrites the old gradient.
3. Every optimizer's `zero_grad()` only resets parameters where `get_requires_grad() && is_grad_initialized()`:
   - `optimizers/adamw.cpp:61-67`
   - `optimizers/adamw_full_precision.cpp:49-55`
   - `optimizers/adamw_composite.cpp` (`MorehAdamW::zero_grad`, `AdamWComposite::zero_grad`)
   - `optimizers/sgd.cpp:34-40`
   - `optimizers/sgd_composite.cpp:30-36`
   - `optimizers/muon_composite.cpp:55-61`

   A frozen parameter is skipped, so its old gradient survives.

`step()` then checks only `is_grad_initialized()` (`adamw.cpp:79-81`, `sgd.cpp:49-52`, `sgd_composite.cpp:46-48`, `muon_composite.cpp:71-73`), finds the stale gradient, and applies it again.

For comparison, `torch.optim.Optimizer.zero_grad` resets `p.grad` for every parameter in `param_groups` with no `requires_grad` check, and Adam and SGD skip any parameter whose `grad` is `None`.

### Expected

After `set_requires_grad(false)`, `step()` leaves the parameter and its optimizer state unchanged, as torch.optim does.

### Suggested fix

The smallest change is to have `Tensor::set_requires_grad(false)` also reset the gradient (`m_grad = ttnn::Tensor();`). Every `step()` already skips a parameter whose gradient is not initialized, so this fixes all seven optimizers in one place. The fix-check block in the script shows the drift going to zero with this change.

The alternative is to drop the `get_requires_grad()` check from each `zero_grad()`, and for a frozen parameter set an empty tensor (as AdamW's `zero_grad` already does, `adamw.cpp:64`) rather than `zeros_like`. With `zeros_like`, SGD and Muon would still apply their momentum buffer to the frozen parameter.

### 1. Steps (exact commands)

1. Download the attached `tt_stale_grad_after_freeze_repro.py`.
2. Run `python3 tt_stale_grad_after_freeze_repro.py` (needs only numpy).

The attached `tt_stale_grad_after_freeze_repro.py` (numpy only) copies this control flow: `add_grad` skips when frozen, `zero_grad` has the `requires_grad` filter, and `step` skips uninitialized gradients. It runs that next to torch semantics on two parameters and freezes B after some steps. It doesn't call ttml, so treat it as a check of the logic rather than a hardware run.

A gtest along the lines of `SGDLateMomentumTest` (sketch, not compiled):

```cpp
class FrozenParameterTest : public ::testing::Test {
public:
    static void SetUpTestSuite() {
        ttml::autograd::ctx().open_device();
    }
    static void TearDownTestSuite() {
        ttml::autograd::ctx().close_device();
    }

protected:
    void TearDown() override {
        ttml::autograd::ctx().reset_graph();
    }
};

// Trains one step, freezes the parameter, then runs the normal loop for a few steps.
// A frozen parameter must not change.
template <typename Optimizer, typename Config>
static void run_frozen_param_case(const Config& cfg) {
    using namespace ttml;

    const std::array<std::size_t, 4> shape = {1, 1, 32, 32};
    autograd::ctx().set_seed(123U);
    auto& gen = autograd::ctx().get_generator();
    xt::xarray<float> w = test_utils::make_uniform_xarray<float>(shape, -1.0F, 1.0F, gen());
    xt::xarray<float> g = test_utils::make_uniform_xarray<float>(shape, 0.25F, 1.0F, gen());

    auto theta = autograd::create_tensor(to_tt(w), true);
    ttml::serialization::NamedParameters params{{"theta", theta}};
    Optimizer opt(params, cfg);

    opt.zero_grad();
    theta->add_grad(to_tt(g));
    opt.step();

    theta->set_requires_grad(false);
    auto frozen = ttml::core::to_xtensor(theta->get_value());
    for (int i = 0; i < 3; ++i) {
        opt.zero_grad();
        theta->add_grad(to_tt(g));  // what backward() does; a no-op while frozen
        opt.step();
    }
    auto after = ttml::core::to_xtensor(theta->get_value());
    EXPECT_TRUE(frozen == after) << "a frozen parameter must not change";
}

TEST_F(FrozenParameterTest, AdamWLeavesFrozenParameterUnchanged) {
    ttml::optimizers::AdamWConfig cfg;
    cfg.lr = 1e-2f;
    run_frozen_param_case<ttml::optimizers::AdamW>(cfg);
}

TEST_F(FrozenParameterTest, SGDMomentumLeavesFrozenParameterUnchanged) {
    ttml::optimizers::SGDConfig cfg;
    cfg.lr = 1e-2f;
    cfg.momentum = 0.9f;
    run_frozen_param_case<ttml::optimizers::SGD>(cfg);
}
```

The same case applies to AdamWFullPrecision, MorehAdamW, AdamWComposite, SGDComposite and MuonComposite.

### 2. Input data / link or description

None needed. The script builds its own random tensors from a fixed seed. No dataset, model or config.

### 3. Frequency

Every run. The script uses a fixed seed and its output is identical each time.

### 1. Software Versions

OS version: macOS 27.0 (the script is plain Python and doesn't depend on the OS)
Software / firmware release version: tt-metal main at c87cf141a5 (source read, not built). Script run with Python 3.14.7 and numpy 2.5.2.

### 2. Hardware Details

Product: none. Not run on Tenstorrent hardware.
Card/System: none. Found by reading source and reproduced with the attached standalone numpy script.

### Is this a regression?

Unknown

### Logs & Diagnostics

Output of `tt_stale_grad_after_freeze_repro.py`:

```
Sanity 1 (force-green): no freezing, torch and ttml emulations must agree exactly
  adamw max param diff: 0.000e+00
  sgd   max param diff: 0.000e+00

Sanity 2: B frozen before it ever gets a gradient, so there is no stale gradient to reuse
  adamw trained 0 steps, frozen for 1000: expected (torch) B drift after freeze = 0.000e+00   actual (ttml) = 0.000e+00   (|B| at freeze max 1.304e+00)
  sgd   trained 0 steps, frozen for 1000: expected (torch) B drift after freeze = 0.000e+00   actual (ttml) = 0.000e+00   (|B| at freeze max 1.304e+00)

Bug case: B trained, then frozen with set_requires_grad(false)
  adamw trained 1 steps, frozen for 100: expected (torch) B drift after freeze = 0.000e+00   actual (ttml) = 1.000e-01   (|B| at freeze max 1.305e+00)
  adamw trained 100 steps, frozen for 1000: expected (torch) B drift after freeze = 0.000e+00   actual (ttml) = 8.844e-01   (|B| at freeze max 1.319e+00)
  sgd   trained 1 steps, frozen for 100: expected (torch) B drift after freeze = 0.000e+00   actual (ttml) = 2.137e+01   (|B| at freeze max 1.316e+00)
  sgd   trained 100 steps, frozen for 1000: expected (torch) B drift after freeze = 0.000e+00   actual (ttml) = 6.300e+01   (|B| at freeze max 2.652e+00)

Fix check: same bug cases with set_requires_grad(false) also clearing the gradient
  adamw trained 1, frozen 100: B drift after freeze = 0.000e+00
  adamw trained 100, frozen 1000: B drift after freeze = 0.000e+00
  sgd   trained 1, frozen 100: B drift after freeze = 0.000e+00
  sgd   trained 100, frozen 1000: B drift after freeze = 0.000e+00

RESULT: FAIL - 4/4 cases update a frozen parameter that torch.optim leaves unchanged
```

[tt_stale_grad_after_freeze_repro.py](https://github.com/user-attachments/files/32619813/tt_stale_grad_after_freeze_repro.py)

### Priority

None

### Impact

Affected workflows: freezing some layers partway through training when the optimizer was built over all of the model's parameters. The frozen parameter keeps changing, with no error or warning.
